File: ruby/test/test_extism.rb
# frozen_string_literal: true

require 'test_helper'

class TestExtism < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil Extism::VERSION
  end

  def test_plugin_call
    plugin = Extism::Plugin.new(manifest)
    res = JSON.parse(plugin.call('count_vowels', 'this is a test'))
    assert_equal res['count'], 4
    res = JSON.parse(plugin.call('count_vowels', 'this is a test again'))
    assert_equal res['count'], 7
    res = JSON.parse(plugin.call('count_vowels', 'this is a test thrice'))
    assert_equal res['count'], 6
    res = JSON.parse(plugin.call('count_vowels', '🌎hello🌎world🌎'))
    assert_equal res['count'], 3
  end

  def test_can_free_plugin
    plugin = Extism::Plugin.new(manifest)
    _res = plugin.call('count_vowels', 'this is a test')
    plugin.free
    assert_raises(Extism::Error) do
      _res = plugin.call('count_vowels', 'this is a test')
    end
  end

  def test_errors_on_bad_manifest
    assert_raises(Extism::Error) do
      _plugin = Extism::Plugin.new({ not_a_real_manifest: true })
    end
  end

  def test_has_function
    plugin = Extism::Plugin.new(manifest)
    assert plugin.has_function? 'count_vowels'
    refute plugin.has_function? 'i_am_not_a_function'
  end

  def test_errors_on_unknown_function
    plugin = Extism::Plugin.new(manifest)
    assert_raises(Extism::Error) do
      plugin.call('non_existent_function', 'input')
    end
  end

  def test_host_functions
    Extism.set_log_file('stdout', 'info')
    func = proc do |current_plugin, inputs, outputs, user_data|
      input = current_plugin.input_as_bytes(inputs.first)
      current_plugin.return_string(outputs.first, "#{input} #{user_data}")
    end
    f = Extism::Function.new('transform_string', [Extism::ValType::I64], [Extism::ValType::I64], func, 'My User Data')
    plugin = Extism::Plugin.new(host_manifest, [f], true)
    result = plugin.call('reflect_string', 'Hello, World!')
    assert_equal result, 'Hello, World! My User Data'
  end

  private

  def manifest
    {
      wasm: [
        {
          path: File.join(__dir__, '../../wasm/code.wasm')
        }
      ]
    }
  end

  def host_manifest
    {
      wasm: [
        {
          path: File.join(__dir__, '../../wasm/kitchensink.wasm')
        }
      ]
    }
  end
end

File: python/tests/test_extism.py
import unittest
import extism
import hashlib
import json
import time
from threading import Thread
from datetime import datetime, timedelta
from os.path import join, dirname


class TestExtism(unittest.TestCase):
    def test_call_plugin(self):
        plugin = extism.Plugin(self._manifest())
        j = json.loads(plugin.call("count_vowels", "this is a test"))
        self.assertEqual(j["count"], 4)
        j = json.loads(plugin.call("count_vowels", "this is a test again"))
        self.assertEqual(j["count"], 7)
        j = json.loads(plugin.call("count_vowels", "this is a test thrice"))
        self.assertEqual(j["count"], 6)
        j = json.loads(plugin.call("count_vowels", "🌎hello🌎world🌎"))
        self.assertEqual(j["count"], 3)

    def test_function_exists(self):
        plugin = extism.Plugin(self._manifest())
        self.assertTrue(plugin.function_exists("count_vowels"))
        self.assertFalse(plugin.function_exists("i_dont_exist"))

    def test_errors_on_unknown_function(self):
        plugin = extism.Plugin(self._manifest())
        self.assertRaises(
            extism.Error, lambda: plugin.call("i_dont_exist", "someinput")
        )

    def test_can_free_plugin(self):
        plugin = extism.Plugin(self._manifest())
        del plugin

    def test_errors_on_bad_manifest(self):
        self.assertRaises(
            extism.Error, lambda: extism.Plugin({"invalid_manifest": True})
        )

    def test_extism_version(self):
        self.assertIsNotNone(extism.extism_version())

    def test_extism_plugin_timeout(self):
        plugin = extism.Plugin(self._loop_manifest())
        start = datetime.now()
        self.assertRaises(extism.Error, lambda: plugin.call("infinite_loop", b""))
        end = datetime.now()
        self.assertLess(
            end,
            start + timedelta(seconds=1.01),
            "plugin timeout exceeded 1000ms expectation",
        )

    def test_extism_host_function(self):
        @extism.host_fn
        def hello_world(plugin, params, results, user_data):
            offs = plugin.alloc(len(user_data))
            mem = plugin.memory(offs)
            mem[:] = user_data
            results[0].value = offs.offset

        f = [
            extism.Function(
                "hello_world",
                [extism.ValType.I64],
                [extism.ValType.I64],
                hello_world,
                b"test",
            )
        ]
        plugin = extism.Plugin(self._manifest(functions=True), functions=f, wasi=True)
        res = plugin.call("count_vowels", "aaa")
        self.assertEqual(res, b"test")

    def test_extism_plugin_cancel(self):
        plugin = extism.Plugin(self._loop_manifest())
        cancel_handle = plugin.cancel_handle()

        def cancel(handle):
            time.sleep(0.5)
            handle.cancel()

        Thread(target=cancel, args=[cancel_handle]).run()
        self.assertRaises(extism.Error, lambda: plugin.call("infinite_loop", b""))

    def _manifest(self, functions=False):
        wasm = self._count_vowels_wasm(functions)
        hash = hashlib.sha256(wasm).hexdigest()
        return {"wasm": [{"data": wasm, "hash": hash}]}

    def _loop_manifest(self):
        wasm = self._infinite_loop_wasm()
        hash = hashlib.sha256(wasm).hexdigest()
        return {
            "wasm": [{"data": wasm, "hash": hash}],
            "timeout_ms": 1000,
        }

    def _count_vowels_wasm(self, functions=False):
        return read_test_wasm("code.wasm" if not functions else "code-functions.wasm")

    def _infinite_loop_wasm(self):
        return read_test_wasm("loop.wasm")


def read_test_wasm(p):
    path = join(dirname(__file__), "..", "..", "wasm", p)
    with open(path, "rb") as wasm_file:
        return wasm_file.read()
